core/modules/file/file.module

```diff
@@ -2789,101 +2789,68 @@ function file_get_type($file) {
 /**
  * Determines if a user may perform the given operation on the specified file.
  *
- * @param $op
+ * @param string $op
  *   The operation to be performed on the file. Possible values are:
+ *   - "create"
  *   - "view"
  *   - "download"
  *   - "update"
  *   - "delete"
- *   - "create"
- * @param $file
+ * @param object|string|NULL $file
  *   The file object on which the operation is to be performed, or file type
- *   (e.g. 'image') for "create" operation.
- * @param $account
+ *   (e.g. 'image').
+ * @param object $account
  *   Optional, a user object representing the user for whom the operation is to
  *   be performed. Determines access for a user other than the current user.
  *
- * @return
+ * @return bool
  *   TRUE if the operation may be performed, FALSE otherwise.
  */
 function file_access($op, $file = NULL, $account = NULL) {
-  $rights = &backdrop_static(__FUNCTION__, array());
-
-  if (!$file && !in_array($op, array('view', 'download', 'update', 'delete', 'create'), TRUE)) {
-    // If there was no file to check against, and the $op was not one of the
-    // supported ones, we return access denied.
+  // Filter out unsupported operations.
+  if (!in_array($op, array('create', 'view', 'download', 'update', 'delete'), TRUE)) {
     return FALSE;
   }
 
-  // If no user object is supplied, the access check is for the current user.
-  if (empty($account)) {
-    $account = $GLOBALS['user'];
-  }
-
-  // $file may be either an object or a file type. Since file types cannot be
-  // an integer, use either fid or type as the static cache id.
-  $cache_id = NULL;
-  if (!empty($file->id)) {
-    $cache_id = $file->id;
-  }
-  elseif (is_string($file) && $op == 'create') {
-    $cache_id = $file;
-  }
-  else {
-    $cache_id = backdrop_hash_base64(serialize($file));
-  }
-
-  // If we've already checked access for this file, user and op, return from
-  // cache.
-  if (isset($rights[$account->uid][$cache_id][$op])) {
-    return $rights[$account->uid][$cache_id][$op];
-  }
+  // If no $file is given or it's only the file type.
+  if (!$file || is_string($file)) {
+    // Filter out unsupported file types.
+    if (is_string($file) && !in_array($file, file_permissions_get_configured_types())) {
+      return FALSE;
+    }
 
-  if (user_access('bypass file access', $account)) {
-    return $rights[$account->uid][$cache_id][$op] = TRUE;
-  }
+    // No restrictions for this account.
```

Comment:
`file_access()` and `$file->access() should be interchangeable. Except `file_access()` has some permission checks that aren't in the class. Perhaps these should be moved to the hook, much like the node entity.

Reply:
To the hook? Sorry, I'm confused. Mind to explain with more details?

Reply:
I'll try to be clearer. Let's see if I'm making sense: I think we should follow more closely the pattern in `node_access()` where it is basically a "wrapper" that decides if `Node::createAccess()` or `$node->access()` is necessary. And we should create the `createAccess()` method on the `File` class so we can at least have the rudimentary checking there. 

Unless I'm missing something the only time we'd be doing a file access check without a file instance is when creating a file.

Reply:
> And we should create the createAccess() method on the File class

Yes, of course we should. That's (currently) a different issue: [#5479](https://github.com/backdrop/backdrop-issues/issues/5479)

> Unless I'm missing something the only time we'd be doing a file access check without a file instance is when creating a file.

That's not fully true, though. Removing the ability to check without instance and/or type would be a regression compared to D7 - where this was possible with Entity API.

Reply:
I'll try to provide more comment soon but just wanted to say I looked at the access callbacks for File entities in the D7 Entity module and in D7 File Entity module. Both of them required either a entity instance or a string of the type. Unless I missed something.

Reply:
@indigoxela might be good to get @quicksketch take on this since we don't agree on the approach here. Maybe add it to next meeting

Reply:
> Maybe add it to next meeting

I tried that twice already, but he didn't show up here. :wink: 

We're a smart team, I'm sure we can discuss this properly and we'll find a solution. Don't you think so? Or maybe if YOU try to add it to the meeting?

Reply:
I originally agreed with @indigoxela that it was a good idea to be able to check for operations other than `create` without having a file object. I'm not so sure that's necessary now. I think the only scenario when we may not get an object is when the op is `create`. 

(I believe I supported the above because Rules needs to do a general access check on entities to see if a user can update, delete etc specific entities without having an actual instance - but from what I remember Rules uses Entity Plus and its own hook implementations to do this, so it may not be necessary in core - I may be wrong; I need to check Rules again). 

I'm also looking at Drupal 7's File Entity module (which contains `file_entity_access()`) and also and Entity API (which contains `entity_access()`) and I don't see the option of checking without an object for ops other than `create`. Am I missing this, @indigoxela? (you said it would be a regression if we took out this option). 

Regarding @herbdool's comment about making `file_access()` a wrapper much in the same way as `node_access()` is - that may be a good idea. He suggested moving all additional checks into the "hook" - meaning moving them into`file_file_access()` (which implements hook_file_access() and gets called from `File::access()`)

Reply:
@argiepiano currently (before this PR) the $file param in function file_access() is optional. Changing that needs the whole process of deprecation and will need a lot of time. That would be OK, if there weren't that urgent problem with an unusable method - which is what I'm trying to fix.

I moved existing code around and added some additional checks for permissions, that weren't considered yet. I did not rewrite the function/method beyond that.

Reply:
I think I wasn't clear. I was not proposing making $file required.

Reply:
Many thanks for clarifying. :+1: